From 1bb0a7e256b09bff7eac61cbeb7c34bf9ca0d2dd Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Tue, 2 Sep 2014 10:15:02 -0400
Subject: [net] fix setting csum_start in skb_segment()

Message-id: <1409652920-431-40-git-send-email-fwestpha@redhat.com>
Patchwork-id: 90177
O-Subject: [PATCH 39/57] [RHEL7.1 net] net: fix setting csum_start in skb_segment()
Bugzilla: 1131999
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Vlad Yasevich <vyasevic@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: 1131999

commit de843723f9b989178762196fb24dd050cbe20ca3
Author: Tom Herbert <therbert@google.com>
Date:   Wed Jun 25 12:51:01 2014 -0700

    net: fix setting csum_start in skb_segment()

    Dave Jones reported that a crash is occurring in

    csum_partial
    tcp_gso_segment
    inet_gso_segment
    ? update_dl_migration
    skb_mac_gso_segment
    __skb_gso_segment
    dev_hard_start_xmit
    sch_direct_xmit
    __dev_queue_xmit
    ? dev_hard_start_xmit
    dev_queue_xmit
    ip_finish_output
    ? ip_output
    ip_output
    ip_forward_finish
    ip_forward
    ip_rcv_finish
    ip_rcv
    __netif_receive_skb_core
    ? __netif_receive_skb_core
    ? trace_hardirqs_on
    __netif_receive_skb
    netif_receive_skb_internal
    napi_gro_complete
    ? napi_gro_complete
    dev_gro_receive
    ? dev_gro_receive
    napi_gro_receive

    It looks like a likely culprit is that SKB_GSO_CB()->csum_start is
    not set correctly when doing non-scatter gather. We are using
    offset as opposed to doffset.

    Reported-by: Dave Jones <davej@redhat.com>
    Tested-by: Dave Jones <davej@redhat.com>
    Signed-off-by: Tom Herbert <therbert@google.com>
    Signed-off-by: Eric Dumazet <edumazet@google.com>
    Fixes: 7e2b10c1e52ca ("net: Support for multiple checksums with gso")
    Acked-by: Tom Herbert <therbert@google.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/core/skbuff.c b/net/core/skbuff.c
index b7bd946..c4d22dd 100644
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -2980,7 +2980,7 @@ struct sk_buff *skb_segment(struct sk_buff *head_skb,
            skb_put(nskb, len),
            len, 0);
    SKB_GSO_CB(nskb)->csum_start =
-       skb_headroom(nskb) + offset;
+       skb_headroom(nskb) + doffset;
    continue;
   }
 
-- 
1.7.1